documentation/projects/proposals/rekognition_data/20240423-implementation_plan_machine_generated_tags_api.md

```diff
@@ -0,0 +1,331 @@
+and
+[`description`](https://github.com/WordPress/openverse/blob/3f3376cfea8a4355229d4fb7d3a69a299de8cba4/ingestion_server/ingestion_server/es_mapping.py#L89)
+fields[^bm25_change]. The `tags` field (specifically the
+[`name` subfield which is what gets indexed](https://github.com/WordPress/openverse/blob/3f3376cfea8a4355229d4fb7d3a69a299de8cba4/ingestion_server/ingestion_server/es_mapping.py#L110-L117))
+does not have an explicit `similarity` defined, and therefore it is already
+using the `BM25` algorithm.
+
+Below is a discussion of each approach, with the advantages & disadvantages for
+taking said approach.
+
+[^bm25_explanation]:
+    Note that it does not do this naively, as Okapi BM25 has additional
+    complexity involved for how prevalent the search term is in the entire
+    dataset. For more information, see: https://en.wikipedia.org/wiki/Okapi_BM25
+
+[^bm25_change]:
+    #751 intends to explore changing this, but it is blocked by #421.
+
+#### Prefer creator-generated tags and exclude machine-generated tags
+
+The easiest way to handle this approach would be to add logic to the data
+refresh's [tag processing step][parse_tags_logic] which would exclude any tags
+that did not match the document's provider. However, this would also mean that
+the machine-generated tags would not be able to contribute to a document's
+scoring (and therefore its search relevancy). This seems counter to our desire
+for adding the machine-generated labels in general and would only serve to
+provide more context/information for API consumers _once they have the result_.
+If they're unable to find the result in the first place, then the presence of
+the machine-generated labels is irrelevant here.
+
+#### Prefer machine-generated tags and exclude creator-generated tags
+
+Similar to
+[the previous step](#prefer-creator-generated-tags-and-exclude-machine-generated-tags),
+this could be accomplished by adding logic to the [tag processing
+step][parse_tags_logic] to exclude tags that have a provider which matches the
+document's provider. This feels like the wrong approach for a few reasons:
+
+- At present, we have far fewer machine-generated tags than we do records with
+  creator-generated tags. This change would necessarily have an impact on search
+  performance, and since tags are one of the 3 text fields that get searched for
+  a query term, it would likely make a large portion of records no longer show
+  up in the results.
+- Many of our sources include
+  [GLAM institutions](<https://en.wikipedia.org/wiki/GLAM_(cultural_heritage)>).
+  These groups often go to great lengths to appropriately catalog and tag the
+  works that we ingest, and excluding those tags from being searched against
+  would be counter to that effort.
+- Machine-generated labeling is inherently biased, and may be incorrect in some
+  cases.
+
+#### Prefer the creator-generated tags, but boost on duplicate machine-generated tags
```

Comment:
This is an interesting option that had never occurred to me, and in some ways I really like the idea behind this approach. It is basically inline with what already exists with the authority boost. However, here we are applying it on the level of an individual piece of metadata based not on the provider of the work, but on a cross-reference with some other source of metadata that corroborates existing metadata.

If we established a heuristic to apply confident _at all_ to a specific piece of metadata, in this case based on a cross-reference with an additional source of metadata, then it's absolutely a generalise approach we can continue to build on and probably make some astounding and meaningful improvements to search relevancy and metadata handling.

That said, it's a missed opportunity for machine tags to not contribute to the discoverability of works that are missing subject based tags at all (like your "best friend" example).

It is probably out of scope of this project to do a combined version, where we use the ranked features to boost the confidence/authority of provider tags based on similarity to machine generated ones, while also incorporating machine tags when they didn't already exist in the provider tags.

I wonder what the volume of completely new tags vs boosted tags would be.

We wouldn't lose the opportunity to switch to this later though, so I don't think it's worth the additional effort (which looks so significant compared to the approach you landed on of basically making no changes to the index, brilliant). We can always revisit a thing like this in the future (e.g., if we incorporated a way for users to "verify" metadata, or more proximately if we incorporate search->result frequency for a work utilising the context of the query that led to that result being clicked to boost the metadata of that result _that was relevant for the query_ rather than the result wholesale... along the lines of quality vs relevancy there).